### PR TITLE
ToC as virtual lines

### PR DESCRIPTION
**Pull request by mrossinek**
Potentially supersedes (the currently broken) #88 

Peek_2021-10-20_23-25

Known issues:
- [ ] cannot navigate onto virtual lines (see [comment here](https://github.com/neovim/neovim/pull/15351#issuecomment-946099466)) [nothing we can do about this for now]
- [x] line joining after trailing modifier does not remove said modifier and following indent
- [x] heading icons are not respected (will probably be replaced by list of links anyways)
- [x] heading seems unnecessary now that the ToC gets rendered right below the insertion

Mixing features:
- [x] support extracting ToC into location/quickfix list
- [x] do we still want a sidebar/split view? (selection popup in previous PR is pretty much broken)

### CODE REVIEW

**Comment by mrossinek** [comment]
Alright I think it's about time for a short update...

This is based on a very early stage and experimental feature of Neovim 0.6
As such, this will definitely not be enabled by default for now.
However, the ability to also open a ToC in a split and to populate the quickfix or location list is something rather useful imo.
As such, this PR should add some examples + documentation explaining how to use this.

So here are some open todos before merging this:
- [ ] add docs/examples
- [ ] refactor to new link syntax format (depends on #160)

**Comment by danymat** [comment]
We should push this feature (#131) to completion in order for the split Toc be really useful.

**Comment by danymat** [comment]
I found out that calling `display_toc` multiple times does not delete the previous virtual text. And scrolling in the buffer does not delete the virtual text too

**Comment by mrossinek** [comment]
> I found out that calling `display_toc` multiple times does not delete the previous virtual text. And scrolling in the buffer does not delete the virtual text too

Good catch, thanks! I will try to update this PR later this week :+1:

**Comment by max397574** [comment]
This PR is quite stable, this is what's left to do:

- [-] Remove virtual texts on cursor move
- [x] Do not duplicate virtual texts if one is already in place
- [ ] Add the ability to follow the link from the TOC split (cc @vhyrro ?)

**Comment by danymat** [comment]
> Second one done (what should I do when user tries to open a second one? sth like log.warn("ToC already displayed")?)

I think this is a great idea 
> 
> I'm not sure if I've understood the third one correctly
> You want that you have the toc in a split, press on a link and then jump to the heading in the normal file?

Yes, exactly (closing the toc buffer too )

**Comment by mrossinek** [comment]
> Remove virtual texts on cursor move

Why would you want to do this?

>> Do not duplicate virtual texts if one is already in place
>
> what should I do when user tries to open a second one? sth like log.warn("ToC already displayed")?

I would not log a warning. Rather jump back to where the ToC is displayed and don't redo the work.

> Yes, exactly (closing the toc buffer too)

Whether or not to close the buffer needs to be configurable.

**Comment by max397574** [comment]
> > > Do not duplicate virtual texts if one is already in place
> > 
> > 
> > what should I do when user tries to open a second one? sth like log.warn("ToC already displayed")?
> 
> I would not log a warning. Rather jump back to where the ToC is displayed and don't redo the work.

I have an if/else so I can do anything there
I can try to jump to the toc
> 
> > Yes, exactly (closing the toc buffer too)
> 
> Whether or not to close the buffer needs to be configurable.
ok I'll do that

**Comment by danymat** [comment]
> Why would you want to do this?

Oh sorry I though it was a feature (it was like at the beginning of this PR)

**Comment by max397574** [comment]
> > Why would you want to do this?
> 
> Oh sorry I though it was a feature (it was like at the beginning of this PR)

so I won't do it ig

**Comment by danymat** [comment]
Okay guys short update: we can now follow links in TOC split !! 🤗 

I just found a bug by the way: #248 

I will update this a little to be able to keep the split open after jumping (with a custom config)

File: lua/neorg/modules/core/integrations/treesitter/module.lua [hunk 1]
```diff
@@ -804,6 +805,15 @@ module.public = {
             _node = _node:parent()
         end
     end,
+
+    --- Get the root node from a `bufnr`
+    --- @param bufnr number
+    --- @return userdata
+    get_buf_root_node = function(bufnr)
```
Comment (mrossinek) [inline lua/neorg/modules/core/integrations/treesitter/module.lua:812]: There is already `get_document_root(bufnr)` for this purpose in this same module.

File: lua/neorg/modules/core/keybinds/default_keybinds.lua [hunk 2]
```diff
@@ -60,6 +60,19 @@ module.public = {
                 silent = true,
                 noremap = true,
             })
+            keybinds.map_event_to_mode("toc-split", {
```
Comment (mrossinek) [inline lua/neorg/modules/core/keybinds/default_keybinds.lua:63]: Just adding a comment here in order to comment on the keybinds file: 
I think we need a keybind for the following calls:
- `:lua neorg.modules.get_module("core.norg.qol.toc").display_toc()` -> displays ToC as non-navigable virtual lines
- `:lua neorg.modules.get_module("core.norg.qol.toc").display_toc(true)` -> displays ToC in split
- `:lua neorg.modules.get_module("core.norg.qol.toc").toqflist(false)` -> adds ToC to quickfix list (if bool is true then location list is used, no need for an extra default keybind imo)

These obviously need to be added to normal mode.

File: lua/neorg/modules/core/norg/esupports/hop/module.lua [hunk 3]
```diff
@@ -32,6 +32,116 @@ module.config.public = {
 }
 
 module.public = {
+    --- Follow link from a specific node
+    --- @param node userdata
+    --- @param split string|nil if not nil, will open a new split with the split mode defined (vsplitr...)
```
Comment (mrossinek) [inline lua/neorg/modules/core/norg/esupports/hop/module.lua:37]: `parsed_link` is not documented. What is it exactly?

File: lua/neorg/modules/core/norg/qol/toc/module.lua [hunk 4]
```diff
@@ -0,0 +1,376 @@
+--[[
+    File: Qol-Toc
+    Title: ToC Generation
+    Summary: Generates a Table of Content from the Neorg file.
+    ---
+In order to use this feature, just add a `= TOC` at some place, and call:
+
+- `lua neorg.modules.get_module('core.norg.qol.toc').display_toc()`
+
+Calling display_toc with true (`...display_toc(true)`) will generate a split view of the Toc.
```
Comment (mrossinek) [inline lua/neorg/modules/core/norg/qol/toc/module.lua:10]: Should also document the `.toqflist()` method

Comment (mrossinek) [reply lua/neorg/modules/core/norg/qol/toc/module.lua:10]: We should also probably add a `Configuration` section

File: lua/neorg/modules/core/norg/qol/toc/module.lua [hunk 5]
```diff
@@ -0,0 +1,376 @@
...
+    find_toc = function()
+        -- Extract any insertion that has a ToC value in it
+        local query = vim.treesitter.parse_query(
+            "norg",
+            [[
+            (insertion
+            (insertion_prefix)
+            item: (capitalized_word) @item
+            parameters: (paragraph_segment)? @parameters
+            (#match? @item "^[tT][oO][cC]$")
```
Comment (mrossinek) [inline lua/neorg/modules/core/norg/qol/toc/module.lua:93]:
```suggestion
            (#match? @item "^T[oO][cC]$")
```
Needs to be capitalized according to the current NFF spec (insertions are capitalized, variables are lowercase)

File: lua/neorg/modules/core/norg/qol/toc/module.lua [hunk 6]
```diff
@@ -0,0 +1,376 @@
...
+        -- All captures are looped over here
+        for id, node in query:iter_captures(root, 0) do
+            -- Extract the name of the capture from the captures table (this makes it easier to perform comparisons on)
+            local capture = query.captures[id]
+
+            -- If the capture name is "item" then we set the item variable inside of node_data
+            -- It can also optionally be "parameters", in which case that variable will be set too
+            node_data[capture] = node
+
+            if node_data.line == nil then
```
Comment (mrossinek) [inline lua/neorg/modules/core/norg/qol/toc/module.lua:121]: Kind of a weird edge case: this currently prevents from two `= Toc` in the same document. Not sure if we ever wanted this, but I'd figure I should point it out.

File: lua/neorg/modules/core/norg/qol/toc/module.lua [hunk 7]
```diff
@@ -0,0 +1,376 @@
...
+                    local line, _, _ = node:start()
+                    local heading_text_node = ts.get_first_node("paragraph_segment", 0, node)
+                    local heading_text = ts_utils.get_node_text(heading_text_node, 0)
+
+                    local prefix = string.rep(display_as_links and "-" or "*", heading_level)
+                        .. (display_as_links and "> " or " ")
+                    local text = prefix
+                        .. (function()
+                            if display_as_links then
+                                return "{# " .. table.concat(heading_text, "") .. "}"
```
Comment (mrossinek) [inline lua/neorg/modules/core/norg/qol/toc/module.lua:189]: Why are you using the magic link target here? What speaks against preserving the heading_level information? We're computing it anyways already and if you use the level here too, you will get the corresponding highlighting in the links, too.

File: lua/neorg/modules/core/queries/native/module.lua [hunk 8]
```diff
@@ -106,7 +106,7 @@ module.public = {
     --- @param bufnr number
     --- @return table
     query_nodes_from_buf = function(tree, bufnr)
-        local root_node = module.private.get_buf_root_node(bufnr)
+        local root_node = module.required["core.integrations.treesitter"].get_buf_root_node(bufnr)
```
Comment (mrossinek) [inline lua/neorg/modules/core/queries/native/module.lua:109]:
```suggestion
        local root_node = module.required["core.integrations.treesitter"].get_document_root(bufnr)
```

**Review by mrossinek — COMMENTED** [review]
Just some observations. Other than that this is shaping up quite nicely :+1:

File: lua/neorg/modules/core/norg/qol/toc/module.lua (hunk 4 above)
Comment (danymat) [reply lua/neorg/modules/core/norg/qol/toc/module.lua:10]: The configuration section is automatically generated (wiki docgen) with the comments on top of each `module.config.public` parameter

File: lua/neorg/modules/core/norg/qol/toc/module.lua (hunk 6 above)
Comment (danymat) [reply lua/neorg/modules/core/norg/qol/toc/module.lua:121]: I think this is a good idea to only allow one `= Toc` in the document. What's the point of allowing multiple ones ?

**Comment by danymat** [comment]
Okay this feature is quite ready to be merged. We still need to:

- [ ] Create a command called `Neorg toc close` that will close the split, or remove the inline toc.
- [ ] Change links to headings in toc split
- [ ] Fix: if custom name in toc (`= TOC custom name`), the custom name is not showing in inline

**Review by danymat — APPROVED** [review]

**Review by vhyrro — APPROVED** [review]

**Pull request merged**